File: Board.py
import random

class Board:
    def __init__(self, dim, points = None):
        self.dim = dim
        # if no points were supplied, create new points
        points = self.create_points() if not points else points
        # check if no point lays outside of the board
        points_check = [all([el <= dim - 1 for el in point]) for point in points]
        assert all(points_check), "Points exceed dimension"
        self.points = points

    def __repr__(self):
        """
        Create a string representation of the matrix
        :return: string
        """
        str = f"Board({self.dim}, points={self.points})"
        return str

    def __str__(self):
        """
        Creates a matrix representation of the board
        """
        matrix = []
        for x in range(self.dim):
            row = []
            for y in range(self.dim):
                val = "X" if (x, y) in self.points else " "
                row.append(val)
            matrix.append(" ".join(row))
        return "\n".join(matrix)

    def __eq__(self, other):
        return id(self) == id(other)

    def create_points(self):
        """
        Used when no points are supplied
        :return: a list of points needed in the init
        """
        # create a half empty matrix
        locs = [(x, y) for x in range(self.dim) for y in range(self.dim)]
        points = []
        for x, y in locs:
            loc_ind = x * self.dim + y
            if loc_ind >= ((self.dim ** 2) / 2):
                points.append((x, y))
        return points

    def neighbor(self, point, direction):
        """
        Get the neighbor of a point in a given direction. Checks whether the point can be occupied
        :param point:
        :param direction:
        :return:
        """
        col, row = point

        if "U" in direction:
            col -= 1
        elif "D" in direction:
            col += 1

        if "R" in direction:
            row += 1
        elif "L" in direction:
            row -= 1

        row_inrange = 0 <= row < self.dim
        col_inrange = 0 <= col < self.dim
        if not all([row_inrange, col_inrange]):
            return None
        if (col, row) in self.points:
            return None
        return (col, row)

    def follow(self, gravity = "DR"):
        assert gravity in ["U", "D", "L", "R", "UL", "UR", "DR", "DL"], "Invalid gravity"
        if len(self.points) == 0:
            return self, False

        point = self.points[0]
        # check if the point can fall down to gravity
        gravpoint = self.neighbor(point, gravity)

        if gravpoint:
            self.points.remove(point)
            self.points.append(gravpoint)
            follow_state = True
            return self, follow_state

        # the point cannot fall down
        # let the point fall to a side
        side_dirs = self.side_directions(gravity)
        side_points = []
        for dir in side_dirs:
            p = self.neighbor(point, dir)
            if p:
                side_points.append(p)

        if side_points != []:
            self.points.remove(point)
            self.points.append(random.choice(side_points))
            follow_state = True
            return self, follow_state

        # nothing happend, put the point back
        self.points.remove(point)
        self.points.append(point)
        # check if any point can follow in the near future
        follow_states = False
        dirs = [gravity] + side_dirs
        for dir in dirs:
            follow_states = any([self.neighbor(p, dir) for p in self.points])
            if follow_states:
                break


        return self, follow_states

    def side_directions(self, direction):
        """
        Returns the both wanted sides
        :param direction:
        :return: list
        """
        # if a combined dir has been provided, return the two main dirs
        if len(direction) == 2:
            return list(direction)
        # else, we need to find the sides of a main dir
        horizontal = ["L", "R"]
        vertical = ["U", "D"]
        if direction in horizontal:
            return [f"{v}{direction}" for v in vertical]
        return [f"{direction}{h}" for h in horizontal]

File: Board.cpp
#include <iostream>
#include <tuple>
#include <vector>
#include <string>
#include <algorithm>
#include "Board.h"

typedef std::tuple<int, int> point;

std::string point_to_tuple(point p)
{
    int p1 = std::get<0>(p);
    int p2 = std::get<1>(p);
    std::string response;
    response += "(";
    response += std::to_string(p1);
    response += ", ";
    response += std::to_string(p2);
    response += ")";
    return response;
};

std::vector<std::string> side_dirs(std::string direction)
{
    std::vector<std::string> dirs;
    // if the direction is one of the corners, length == 2, so return both main dirs
    if (direction.size() == 2)
    {
        for (int i = 0; i < direction.size(); i++)
        {
            std::string new_element(1, direction[i]);
            dirs.push_back(new_element);
        }
        return dirs;
    }
    // we received a main dir, so give back the according corner dirs
    if (direction == "L" || direction == "R")
    {
        dirs.push_back("U" + direction);
        dirs.push_back("D" + direction);
    }
    else
    {
        dirs.push_back(direction + "L");
        dirs.push_back(direction + "R");
    }
    return dirs;
}

Board::Board(int dim, std::vector<point> points)
{
    this->dim = dim;
    this->points = points;
}

bool Board::poi_in_points(point quest_point)
{
    for (point point_in_points : points)
    {
        if (point_in_points == quest_point)
        {
            return true;
        }
    }
    return false;
}

void Board::print_points()
{
    for (point p : points)
    {
        point_to_tuple(p);
    }
    std::cout << std::endl;
}

void Board::print_board()
{
    std::string board_rows = "";
    for (int col = 0; col < dim; col++)
    {
        std::string row_string = "";
        for (int row = 0; row < dim; row++)
        {
            // create a point based on the for loops
            point point(col, row);
            // check if point in points
            bool found_point = std::find(points.begin(), points.end(), point) != points.end();
            row_string += found_point ? "X" : "-";
            row_string += " ";
        }
        // Append row to the board string
        board_rows += row_string;
        board_rows += "\n";
    }
    // print board
    std::cout << board_rows;
}

point Board::neighbor(point old_point, std::string direction)
{
    int col = std::get<0>(old_point);
    int row = std::get<1>(old_point);
    if (direction.find('U') != std::string::npos)
    {
        col--;
    }
    else if (direction.find('D') != std::string::npos)
    {
        col++;
    }

    if (direction.find('R') != std::string::npos)
    {
        row++;
    }
    else if (direction.find('L') != std::string::npos)
    {
        row--;
    }

    point unfound_point(-1, -1); // init an impossible point
    // check if any points exceeds dimension
    if (row == -1 || row == dim || col == -1 || col == dim)
    {
        return unfound_point;
    }
    // check if neighbor is already in points
    point quest_point(col, row);
    if (this->poi_in_points(quest_point))
    {
        return unfound_point;
    }
    return quest_point;
}

bool Board::follow_gravity(std::string gravity = "DR")
{
    for (int i = 0; i < points.size(); i++)
    {
        // get the first point in points
        point poi = points[i];
        point unfound_point(-1, -1);
        // calculate the gravpoint
        point gravpoint(this->neighbor(poi, gravity));
        // if there is a gravpoint, push it to the list
        if (gravpoint == unfound_point)
        {
            continue;
        }
        points.erase(points.begin() + i);
        points.push_back(gravpoint);
        return true; // the follow worked
    }
    return false;
}

bool Board::follow_side(std::string gravity = "DR")
{
    std::vector<std::string> side_directions(side_dirs(gravity));
    point unfound_point(-1, -1);
    for (int i = 0; i < points.size(); i++)
    {
        // get the first point in points
        point poi = points[i];
        // calculate both sides
        std::vector<point> side_points;
        for (std::string dir : side_directions)
        {
            point side_point(this->neighbor(poi, dir));
            if (side_point == unfound_point)
            {
                continue;
            }
            side_points.push_back(side_point);
        }
        // len(side_points) == 0 => there weren't any side points for this poi
        if (side_points.size() == 0)
        {
            continue;
        }
        else if (side_points.size() == 1)
        {
            points.erase(points.begin() + i);
            points.push_back(side_points[0]);
            return true;
        }
        int index = rand() % 2; // get a random number between 0 and 1
        points.erase(points.begin() + i);
        points.push_back(side_points[index]);
        return true;
    }
    return false;
}

bool Board::follow(std::string gravity = "DR")
{
    if (points.size() == 0)
    {
        return false;
    }
    bool follow_grav = this->follow_gravity(gravity);
    if (follow_grav)
    {
        return true;
    }
    // no points can follow to gravity
    // check if a point can fall to the side
    return this->follow_side(gravity);
}
